internal/subscriber/retry/delayconsumer.go

```diff
@@ -0,0 +1,186 @@
+				dc.resume()
+			}
+
+			resp, err := dc.consumer.ReceiveMessages(dc.ctx, messagebroker.GetMessagesFromTopicRequest{NumOfMessages: 10, TimeoutMs: int(defaultBrokerOperationsTimeoutMs)})
+			if err != nil {
+				if !messagebroker.IsErrorRecoverable(err) {
+					logger.Ctx(dc.ctx).Errorw("delay-consumer: error in receiving messages", dc.LogFields("error", err.Error())...)
+					return
+				}
+			}
+
+			for _, msg := range resp.PartitionOffsetWithMessages {
+				dc.processMsg(msg)
+			}
+		}
+	}
+}
+
+// resumes a paused delay-consumer. Additionally process any previously cached messages in buffer.
+func (dc *DelayConsumer) resume() {
+	logger.Ctx(dc.ctx).Infow("delay-consumer: resuming", dc.LogFields("error", nil)...)
+	dc.consumer.Resume(dc.ctx, messagebroker.ResumeOnTopicRequest{Topic: dc.topic})
+	dc.isPaused = false
+
+	// process the cached message as well
+	dc.processMsg(*dc.cachedMsg)
+	dc.cachedMsg = nil
+}
+
+// paused an active delay-consumer. Additionally caches the last seen message locally.
+func (dc *DelayConsumer) pause(msg *messagebroker.ReceivedMessage) {
+	logger.Ctx(dc.ctx).Infow("delay-consumer: pausing", dc.LogFields()...)
+	dc.consumer.Pause(dc.ctx, messagebroker.PauseOnTopicRequest{Topic: dc.topic})
+	dc.isPaused = true
+	dc.cachedMsg = msg
+}
+
+// push a message onto the configured dead letter topic
+func (dc *DelayConsumer) pushToDeadLetter(msg *messagebroker.ReceivedMessage) error {
+	logger.Ctx(dc.ctx).Infow("delay-consumer: pushing to dead-letter", dc.LogFields()...)
+	dlProducer, err := dc.bs.GetProducer(dc.ctx, messagebroker.ProducerClientOptions{
+		Topic:     msg.DeadLetterTopic,
+		TimeoutMs: defaultBrokerOperationsTimeoutMs,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = dlProducer.SendMessage(dc.ctx, messagebroker.SendMessageToTopicRequest{
+		Topic:         msg.DeadLetterTopic,
+		Message:       msg.Data,
+		TimeoutMs:     int(defaultBrokerOperationsTimeoutMs),
+		MessageHeader: msg.MessageHeader,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// processes a given message of the delay-topic. takes care of orchestrating the checks to determine pause,resume of the consumer,
+// push to dead-letter topic in case the number of retries breaches allowed threshold.
+func (dc *DelayConsumer) processMsg(msg messagebroker.ReceivedMessage) {
+	if msg.CanProcessMessage() {
+		if dc.isPaused {
+			dc.resume()
+		}
+
+		if msg.HasReachedRetryThreshold() {
+			// push to dead-letter topic directly in such cases
+			err := dc.pushToDeadLetter(&msg)
+			if err != nil {
+				logger.Ctx(dc.ctx).Errorw("delay-consumer: failed to push to dead-letter topic", "topic", msg.DeadLetterTopic, "error", err.Error())
+				return
+			}
+		} else {
+			// submit to handler
+			err := dc.handler.Do(dc.ctx, msg)
+			if err != nil {
+				logger.Ctx(dc.ctx).Errorw("delay-consumer: error in msg handler", dc.LogFields("error", err.Error())...)
+				return
+			}
+		}
+
+		// commit the message
```

Comment:
commit should be taken care by ack, if we commit here we wont be able to retry on failure

Reply:
This is a commit on the delay topic. `ack()` only has access to primary and primary-retry topics.